server.go

```diff
@@ -677,11 +724,34 @@ func (s *Server) processUnaryRPC(t transport.ServerTransport, stream *transport.
 		if trInfo != nil {
 			trInfo.tr.LazyLog(&payload{sent: true, msg: reply}, true)
 		}
-		return t.WriteStatus(stream, statusCode, statusDesc)
+		errWrite := t.WriteStatus(stream, statusCode, statusDesc)
+		if statusCode != codes.OK {
+			return Errorf(statusCode, statusDesc)
+		}
+		return errWrite
 	}
 }
 
 func (s *Server) processStreamingRPC(t transport.ServerTransport, stream *transport.Stream, srv *service, sd *StreamDesc, trInfo *traceInfo) (err error) {
+	if stats.On() {
+		begin := &stats.Begin{
+			BeginTime: time.Now(),
+		}
+		stats.Handle(stream.Context(), begin)
+	}
+	defer func() {
+		if stats.On() {
+			var e error
+			if err != nil && err != io.EOF {
+				e = toRPCErr(err)
+			}
+			end := &stats.End{
+				EndTime: time.Now(),
+				Error:   e,
+			}
+			stats.Handle(stream.Context(), end)
+		}
+	}()
```

Comment:
Can we move this part (from line 736-754 to the caller of processStreamingRPC and processUnaryRPC) since they are common for both cases?